### PR TITLE
Fixed full name generation for locales where names have genders

### PR DESCRIPTION
**Pull request by binarycode**
Previously `Name.FullName()` could generate full name where first name and last name have different genders (like "Дарья Силин" or "Виталий Кудряшева" for Russian locale). That issue is fixed by this PR.

Unfortunately the fix changed the number of calls to `Random` in some situations, and that cause a lot of unrelated tests to break, I had to fix those as well.

### CODE REVIEW

**Comment by bchavez** [comment]
Hi Igor,

Thank you for your Pull Request.

Yeah, the breaking of the unrelated tests is intentional when changes like this occur. The hard-coded test values give me insight into how changes in the source affect the deterministic output of **Bogus**. Ultimately, breaking tests values are a signal for me and others downstream that a major version increment is necessary for the next release.

I haven't got much sleep yet; so I'll try taking a closer look in about 8 hours or so.

:collision: :dizzy: ***["Crashing, hit a wall. Right now I need a miracle..."](https://www.youtube.com/watch?v=s8XIgR5OGJc)***

File: Source/Bogus/DataSets/Name.cs [hunk 1]
```diff
@@ -41,17 +39,18 @@ public Name(string locale = "en") : base(locale)
...
+            if( gender == Gender.Male )
+            {
+               return GetRandomArrayItem("male_first_name");
+            }
+            return GetRandomArrayItem("female_first_name");
          }
-         return GetRandomArrayItem("female_first_name");
+
+         return GetRandomArrayItem("first_name");
       }
```
[the remark this answers is not in the document]
Comment (binarycode) [reply Source/Bogus/DataSets/Name.cs]: Yes, both first name and last name should be of same gender.

I've changed the `Name.FirstName()` code to match `Name.LastName()` logic, as it appears to be more correct to me. Consider a case when we have locale where names do not have genders (like "en" for example), what if someone calls `Name.FirstName(Name.Gender.Male)`?

Here is the snippet before the change:
```
if( gender is null && HasFirstNameList )
   return GetRandomArrayItem("first_name");
```

But gender is not null in this case, so the function will not return item from "first_name" list, it will try to use "male_first_name" or "female_first_name" fields (and fail as this locale does not have those).

That logic was changed to check if we support gendered first name, and if not - use "first_name", else - use appropriate "male_first_name" or "female_first_name" column.

`HasFirstNameList` was removed as it's mutually exclusive with `SupportsGenderFirstNames` - if we don't support gendered first names we must have "first_name" column, or else we don't have any place to take the names from.

[the remark this answers is not in the document]
Comment (bchavez) [reply Source/Bogus/DataSets/Name.cs]: Hi Igor,

Thank you for the clarification. Are you aware of any other locales that have the same gender requirement for first-name and last-name? It would be nice to know of other locales that have this feature as it could provide more clarity in correcting this problem.

To answer your question, I'd like to show that the `en` locale supports gendered `male|female_first_name` and non-gendered `first_name`.

devenv_1724

So, calling `Name.FirstName(Name.Gender.Male)` on the `en` locale should, in fact, get you a male first name.

The more pertinent locale where your `FirstName()` code changes could be an issue is in the `de` German locale. Specifically, the `de` locale contains their own `first_name` set, but doesn't have an explicit set for `male|female_first_name` as shown below:

devenv_1725

Now let's compare the code execution paths and fallback mechanisms between the existing code and your proposed changes in `FirstName` respectively:
```csharp
//FirstName BEFORE changes
public string FirstName(Gender? gender = null)
{
   if( gender is null && HasFirstNameList )
      return GetRandomArrayItem("first_name");

   if( gender is null )
      gender = this.Random.Enum<Gender>();

   if( gender == Gender.Male )
   {
      return GetRandomArrayItem("male_first_name");
   }
   return GetRandomArrayItem("female_first_name");
}

//Your FirstName changes
public string FirstName2(Gender? gender = null)
{
   if (SupportsGenderFirstNames)
   {
      gender = gender ?? this.Random.Enum<Gender>();

      if (gender == Gender.Male)
      {
         return GetRandomArrayItem("male_first_name");
      }
      return GetRandomArrayItem("female_first_name");
   }

   return GetRandomArrayItem("first_name");
}
```
And consider the following invocation code:
```csharp
[Fact]
public void gender_name_test_for_de_locale()
{
   var n = new Name("de");

   //FirstName BEFORE changes
   n.FirstName(); // Delegates to "de: first_name"
   n.FirstName(Name.Gender.Male); // Delegates to "en: male_first_name", FALLBACK to 'en'

   //Your FirstName changes
   n.FirstName2(); // Delegates to "de: first_name"
   n.FirstName2(Name.Gender.Male); // Delegates to "de: first_name"
}
```
The `en` fallback is used when a `de` locale is specified and `n.FirstName(Name.Gender.Male);` is invoked. The invocation would fallback to `en: male_first_name`; whereas your `FirstName2` code changes don't fall back to a male gender name  in the `en`.

The change is subtle but the ***execution paths***, ***behavior***, and ***results***, are in fact different. This is the basis of my main concern about changing the code in the `FirstName()` method. I hope this makes sense; please correct me if I made a mistake in my analysis.

----

Having reviewed the changes in `FirstName`, I think the best approach forward is to revert the changes made in `FirstName` since `FirstName` changes are not relevant to fixing the primary issue at hand. Also, so we don't change the semantic behavior for locales like `de`. Most of the code fixes should be done in `FullName`.

I think the code in `FullName` should look something like:
```csharp
public string FullName(Gender? gender = null)
{
   // 'ru' locale requires a gender to be specified for both first and last name.
   // Gender is not picked when 'en' locale is specified
   // because SupportsGenderLastNames = false when 'en' is specified.
   if( SupportsGenderFirstNames && SupportsGenderLastNames )
      gender = gender ?? this.Random.Enum<Gender>();

   return $"{FirstName(gender)} {LastName(gender)}";
}
```

Basically, with this paritcular gender/name issue, it comes down to submitting a PR with the minimal amount of changes to fix the issue.

Thanks,
Brian

:fallen_leaf: :leaves: **[Distance - Falling (ft Alys Be)](https://www.youtube.com/watch?v=WIfwIv4wNCE)**

[the remark this answers is not in the document]
Comment (binarycode) [reply Source/Bogus/DataSets/Name.cs]: You're right - I should have broken this PR in two, addressing each of my concerns in separate one to reduce confusion.

I've reverted `FirstName()` changes (along with test changes) and added gender support check to `FullName()`.

However I still have some concerns about this logic, please correct me if those should be moved to separate issues/PRs: first of all I'm not sure that falling back to "en" locale is a good idea - let's consider your example:

```csharp
var n = new Name("de");
n.FirstName(Name.Gender.Male); // Delegates to "en: male_first_name", FALLBACK to 'en'
```

User wanted to generate German male name, but instead he got English male name. I think it would be better to give him German non-gendered name.

My second concern is about `LastName()`, here is the current implementation:

```csharp
      public string LastName(Gender? gender = null)
      {
         if( SupportsGenderLastNames )
         {
            gender = gender ?? this.Random.Enum<Gender>();

            if( gender == Gender.Male )
            {
               return GetRandomArrayItem("male_last_name");
            }
            return GetRandomArrayItem("female_last_name");
         }

         return GetRandomArrayItem("last_name");
      }
```

Shouldn't it be changed to

```csharp
public string LastName(Gender? gender = null)
{
   if( gender is null && HasLastNameList )
      return GetRandomArrayItem("last_name");

   if( gender is null )
      gender = this.Random.Enum<Gender>();

   if( gender == Gender.Male )
   {
      return GetRandomArrayItem("male_last_name");
   }
   return GetRandomArrayItem("female_last_name");
}
```

so that it would use the fallback logic you mentioned. The thing is I'm confused why we use the fallback logic in `FirstName()` and do not use it in `LastName()`, as essentialy they do the same thing (the only difference is that they take names from different columns).

Thanks!

[the remark this answers is not in the document]
Comment (bchavez) [reply Source/Bogus/DataSets/Name.cs]: Hi Igor,

Thanks for reverting those changes. Also, thank you for your PR contribution. I'll have a new release with your fixes published shortly.

You're right, you can equally make a compelling argument about whether or not the fallback to `en` mechanism for the German `FirstName()` is appropriate or not. I can't claim to have the right answer here except to say *maybe* "it feels right"... I'll try to elaborate my thoughts:

From an API design standpoint, let's say we're in the German locale and we have the following code:
```csharp
1. var f = new Faker("de");
2. var iCanSeeICanGetAName = f.Name.FirstName();
3. var myIntentHereIsGettingAMaleName = f.Name.FirstName(Name.Gender.Male);
```
Being able to differentiate between gendered names is important enough to have **Bogus** do something "correct" when **Bogus** is in a situation like this. On `Line 3` above, the developer has an explicit intention of getting an **Male** name. Suppose we **did not** fallback to `en`, and instead we returned an item in `de: first_name`. The `de: first_name` list contains ***male*** and ***female*** names.

So, if **Bogus** returned a ***Female*** name when the intent was `Name.Gender.Male`; the developer would likely think **Bogus** is broken because **Bogus** did not do what the developer wanted when the intention was very explicit. I'd have to explain why this API call is not broken and give them a workaround. The only way out of a situation like this is to use a (somewhat) hidden feature in the `Name` data set, which switches to an explicit `en` locale:
```csharp
//This is the way out, force the German developer wanting a male name
//to use 'en' locale
3. var iReallyWantAMaleName = f.Name["en"].FirstName(Name.Gender.Male);
```

---

Now, let's run through the situation where we fallback to `en` with the same code and intent:
```csharp
1. var f = new Faker("de");
2. var iCanSeeICanGetAName = f.Name.FirstName();
3. var myIntentHereIsGettingAMaleName = f.Name.FirstName(Name.Gender.Male);
```
Worst case, on `Line 3` we fallback to `en` and return a **Male** name (guaranteed). At least we've somewhat satisfied the developer's intent. Returning a **Female** name in this API scenario would be, I think, completely orthogonal to the developer's intent.

Let's say, the German developer is not pleased with the `en` fallback of english male names, then their "way out" is to simply back out of the explicit call for a **Male** name and remove the `Name.Gender.Male` parameter. The developer's code is reverted back to it's original API intent:
```csharp
//Developer's source code maps back to original intent.
3. var okayOnlyWayToGetOnlyGermanNames = f.Name.FirstName(); 
```
It's hard to say which way is the "best way". Ultimately, issues like this come down to faker.js' incomplete data sets. I wish I had enough resources to fill the gaps in locale data and correct issues like this but I think this is the best we can do with the data we have.

Also, a minor point is that the current implementation hasn't been a problem so far. However, if it does become a problem in the future, I don't have any qualms revisiting the issue and implementation choices.

I hope these thoughts are reasonable.

Thanks,
Brian

:dancers: :dancer: ***["Ich zeig Dir wie ich tanz... Komm lass uns geh'n..."](https://www.youtube.com/watch?v=yI91jzIvrx4)***

**Comment by bchavez** [comment]
Hi Igor,

Your changes been published in [`Bogus v22.1.3`](https://www.nuget.org/packages/Bogus/22.1.3). :+1: :sunglasses:

Thanks again for your contribution!
Brian

:car: :blue_car: ***["Let the good times roll..."](https://www.youtube.com/watch?v=7BDBzgHXf64)***
